From 402ed8ad910882880e14da68bbbc125cd363662d Mon Sep 17 00:00:00 2001
From: Bo Gan <ganb@vmware.com>
Date: Thu, 10 Nov 2022 10:17:25 +0000
Subject: [PATCH] vfio: Only set INTX_DISABLE bit during disable

Do not touch other PCI command register bits, such as IO/MEMORY enable
bits, during vfio_pci_disable. This avoids decoding getting disabled
and later enabled again in pci_restore_state. For virtualized guests,
enabling/disabling decoding is costly, as the nested pagetable mappings
need updates and get synchronized across all vcpus.

Currently this optimization is effective only with disable_resets=1.

Signed-off-by: Bo Gan <ganb@vmware.com>
---
 drivers/vfio/pci/vfio_pci.c | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/drivers/vfio/pci/vfio_pci.c b/drivers/vfio/pci/vfio_pci.c
index 3fad4e1c99da..3ed1e5e86644 100644
--- a/drivers/vfio/pci/vfio_pci.c
+++ b/drivers/vfio/pci/vfio_pci.c
@@ -384,7 +384,15 @@ static void vfio_pci_disable(struct vfio_pci_device *vdev)
 	 * Disable INTx and MSI, presumably to avoid spurious interrupts
 	 * during reset.  Stolen from pci_reset_function()
 	 */
-	pci_write_config_word(pdev, PCI_COMMAND, PCI_COMMAND_INTX_DISABLE);
+	if (vfio_pci_resets_disabled()) {
+		u16 cmd;
+		pci_read_config_word(pdev, PCI_COMMAND, &cmd);
+		pci_write_config_word(pdev, PCI_COMMAND,
+				      cmd | PCI_COMMAND_INTX_DISABLE);
+	} else {
+		pci_write_config_word(pdev, PCI_COMMAND,
+				      PCI_COMMAND_INTX_DISABLE);
+	}
 
 	/*
 	 * Try to get the locks ourselves to prevent a deadlock. The
-- 
2.25.1

